File: rp/validation/frequency_2_words_validator.py
from validation.base_validator import BaseValidator
from dictionary.two_sequence_dictionary import FrequencyDictionary
from typing import List

class FrequencyValidator(BaseValidator):
    def __init__(self, dictionary: FrequencyDictionary, chunks: List[str]):
        self.dictionary = dictionary
        self.all_chunks = chunks
    
    def validate_chunk(self, chunk1: str, chunk2: str) -> bool:
        ch1 = chunk1.split()
        ch2 = chunk2.split()

        if not ch1 or not ch2: return True
        if len(ch1) < 2 and len(ch2) < 2: return True


        if chunk1[-1] == ' ' or chunk2[0] == ' ':
            return self.dictionary.contains(ch1[-1], ch2[0])

        else:
            new_word = ch1[-1] + ch2[0]
            if new_word not in self.dictionary.data: return False
            if len(ch1) > 2:
                if not self.dictionary.contains(ch1[-2], new_word): return False
            if len(ch2) > 2:
                if not self.dictionary.contains(new_word, ch2[1]): return False
            return True
    
    def validate_text(self, text: str, required_chunks:str) -> bool:
        for chunk in required_chunks:
            if chunk not in text: return False
        return True

File: text/validation/frequency_2_words_validator.cpp
#include "frequency_2_words_validator.hpp"
#include "../text_util.hpp"

bool FrequencyValidator::validate_chunk(const std::string& chunk1, const std::string& chunk2) const {
    std::vector<std::string> ch1 = split_words(chunk1);
    std::vector<std::string> ch2 = split_words(chunk2);

    if (ch1.empty() || ch2.empty()) return true;
    if (ch1.size() < 2 && ch2.size() < 2) return true;

    if (chunk1.back() == ' ' || chunk2.front() == ' ') {
        return dictionary.contains(ch1.back(), ch2.front());
    }

    std::string new_word = ch1.back() + ch2.front();
    if (dictionary.data.count(new_word) == 0) return false;
    if (ch1.size() > 2) {
        if (!dictionary.contains(ch1[ch1.size() - 2], new_word)) return false;
    }
    if (ch2.size() > 2) {
        if (!dictionary.contains(new_word, ch2[1])) return false;
    }
    return true;
}

bool FrequencyValidator::validate_text(const std::string& text, const std::vector<std::string>& required_chunks) const {
    for (const std::string& chunk : required_chunks) {
        if (text.find(chunk) == std::string::npos) return false;
    }
    return true;
}
